File: src/main/java/com/store/dto/OrderResponseDto.java
package com.store.dto;

import com.store.domain.orders.OrderItem;
import com.store.domain.orders.OrderStatus;
import com.store.domain.orders.Orders;
import com.store.domain.product.ProductColor;
import com.store.domain.product.ProductImage;
import com.store.domain.product.ProductOption;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class OrderResponseDto {

    private Long orderId;

    private Long accountId;

    private LocalDateTime orderDate;

    private OrderStatus status;

    private List<OrderItemResponseDto> itemList;

    public static OrderResponseDto from(Orders order) {
        return OrderResponseDto.builder()
                .orderId(order.getId())
                .accountId(order.getAccount().getId())
                .orderDate(order.getOrderDate())
                .status(order.getStatus())
                .itemList(getItemResponseDtos(order))
                .build();
    }

    private static List<OrderItemResponseDto> getItemResponseDtos(Orders order) {
        List<OrderItemResponseDto> itemResponseDtos = new ArrayList<>();
        List<OrderItem> orderItems = order.getOrderItems();
        for (OrderItem item : orderItems) {
            ProductOption option = item.getOption();
            ProductColor color = option.getProductColor();
            itemResponseDtos.add(OrderItemResponseDto.builder()
                    .orderItemId(item.getId())
                    .status(item.getStatus())
                    .quantity(item.getQuantity())
                    .price(item.getPrice())
                    .color(color.getColor())
                    .size(option.getSize())
                    .imageUrl(getMainImageUrl(color))
                    .build());
        }
        return itemResponseDtos;
    }

    private static String getMainImageUrl(ProductColor color) {
        Optional<String> mainImageUrlOpt = color.getImages().stream()
                .filter(ProductImage::isMain)
                .map(ProductImage::getImageUrl)
                .findFirst();
        return mainImageUrlOpt.orElse("");
    }

}

File: src/main/java/com/store/dto/OrderResponseDto.kt
package com.store.dto

import com.store.domain.orders.OrderStatus
import com.store.domain.orders.Orders
import com.store.domain.product.ProductColor
import com.store.domain.product.ProductOption
import java.time.LocalDateTime

data class OrderResponseDto(
    val orderId: Long? = null,
    val accountId: Long? = null,
    val orderDate: LocalDateTime? = null,
    val status: OrderStatus? = null,
    val itemList: List<OrderItemResponseDto> = emptyList(),
) {
    companion object {
        @JvmStatic
        fun from(order: Orders): OrderResponseDto = OrderResponseDto(
            orderId = order.id,
            accountId = order.account.id,
            orderDate = order.orderDate,
            status = order.status,
            itemList = getItemResponseDtos(order),
        )

        private fun getItemResponseDtos(order: Orders): List<OrderItemResponseDto> =
            order.orderItems.map { item ->
                val option: ProductOption = item.option
                val color: ProductColor = option.productColor
                OrderItemResponseDto(
                    orderItemId = item.id,
                    status = item.status,
                    quantity = item.quantity,
                    price = item.price,
                    color = color.color,
                    size = option.size,
                    imageUrl = getMainImageUrl(color),
                )
            }

        private fun getMainImageUrl(color: ProductColor): String =
            color.images.firstOrNull { it.isMain }?.imageUrl ?: ""
    }
}
